src/cmap/connection_pool.ts

```diff
@@ -432,18 +439,72 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
       return;
     }
 
+    const oldGeneration = this[kGeneration];
+
     // handle non load-balanced case
     this[kGeneration] += 1;
     const alreadyPaused = this[kPoolState] === PoolState.paused;
     this[kPoolState] = PoolState.paused;
 
     this.clearMinPoolSizeTimer();
     if (!alreadyPaused) {
-      this.emit(ConnectionPool.CONNECTION_POOL_CLEARED, new ConnectionPoolClearedEvent(this));
+      this.emit(
+        ConnectionPool.CONNECTION_POOL_CLEARED,
+        new ConnectionPoolClearedEvent(this, { interruptInUseConnections })
+      );
     }
+
+    process.nextTick(() =>
+      this.pruneConnections({ minGeneration: oldGeneration, interruptInUseConnections })
+    );
+
     this.processWaitQueue();
   }
 
+  /**
+   * Closes all checked in perished connections in the pool with a resumable PoolClearedOnNetworkError.
+   *
+   * If interruptInUseConnections is `true`, this method attempts to kill checked out connections as well.
+   * Only connections where `connection.generation <= minGeneration` are killed.  Connections are closed with a
+   * resumable PoolClearedOnNetworkTimeoutError.
+   */
+  private pruneConnections({
```

Comment:
why does this method do anything with the connections that aren't checked out? the spec doesn't say anything about aggressively removing stale connections from the pool (in fact, it says the opposite, that they should be "lazily" closed), it's only concerned about **in use** connections and failing any operations that might be using them as quickly as possible. Our logic already ensures that a stale connection will never be handed to an operation, so the logic in lines 478-489 is unnecessary (and also incorrect, since you'd be actually applying an an in use error to connections that are not in use). I suggest renaming this method to some variation of "interruptInUseConnections" (maybe "interruptActiveConnections" to refer to the in use and pending ones), and guard L457-459 with an if check on `interruptInUseConnections`